sourcegraph/sourcegraph
# Charts: [Line Chart] Add better support for stacked series data

Closes https://github.com/sourcegraph/sourcegraph/issues/33349

## Background

| Before  | After |
| ------------- | ------------- |
| <img width="397" alt="Screenshot 2022-04-04 at 14 56 38" src="https://user-images.githubusercontent.com/18492575/161532066-60540676-27c8-4392-ae3e-4825e65efc09.png">  | <img width="385" alt="Screenshot 2022-04-04 at 14 56 54" src="https://user-images.githubusercontent.com/18492575/161532082-202c0081-f3a1-4a9a-92b2-c486785d5759.png">  |


As you can see prior to this PR the line chart had a bad visual and data stacking. If points from two or more series don't share one x-axis coordinate (time) the stacking logic fails and visually we didn't handle this as well, therefore wierd stacked area paths. 

In this PR we handle stacking in series data in the right way. If points from two series don't have the same x time coordinate we're trying to find the closest points from the series on the stack and interpolate values to find the base value for the current point. 

<img width="534" alt="Screenshot 2022-04-04 at 15 06 02" src="https://user-images.githubusercontent.com/18492575/161531308-8eec1f9a-00bd-4f82-9ba9-28b1e222f2c4.png">

## Test plan

- Make sure that all stories work as they worked before 
- Check the performance of the line chart rendering (should be the same or better)
- Check that stacked series work as they should for non-aligned by x-axis points


## App preview:
- [Link](https://sg-web-vk-fix-stacked-line-chart.onrender.com)
